internal/attestation/aws/snp/validator.go

```diff
@@ -9,99 +9,212 @@ package snp
+		log:             log,
+	}
+
 	v.Validator = vtpm.NewValidator(
 		cfg.Measurements,
-		getTrustedKey,
-		v.tpmEnabled,
+		v.getTrustedKey,
+		func(vtpm.AttestationDocument, *attest.MachineState) error { return nil },
 		log,
 	)
-	v.getDescribeClient = getEC2Client
 	return v
 }
 
-// getTrustedKeys return the public area of the provides attestation key.
-// Normally, here the trust of this key should be verified, but currently AWS does not provide this feature.
-func getTrustedKey(_ context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
+// getTrustedKeys return the public area of the provided attestation key (AK).
+// Ideally, the AK should be bound to the TPM via an endorsement key, but currently AWS does not provide one.
+// The AK's digest is written to the SNP report's userdata field during report generation.
+// The AK is trusted if the report can be verified and the AK's digest matches the digest of the AK in attDoc.
+func (v *Validator) getTrustedKey(ctx context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
 	// Copied from https://github.com/edgelesssys/constellation/blob/main/internal/attestation/qemu/validator.go
 	pubArea, err := tpm2.DecodePublic(attDoc.Attestation.AkPub)
 	if err != nil {
-		return nil, err
+		return nil, newDecodeError(err)
+	}
+
+	pubKey, err := pubArea.Key()
+	if err != nil {
+		return nil, fmt.Errorf("getting public key: %w", err)
+	}
+
+	akDigest, err := sha512sum(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("calculating hash of attestation key: %w", err)
+	}
+
+	if err := v.reportValidator.validate(ctx, attDoc, v.ask, v.ark, akDigest, v.cfg, v.log); err != nil {
+		return nil, fmt.Errorf("validating SNP report: %w", err)
 	}
 
 	return pubArea.Key()
 }
 
-// tpmEnabled verifies if the virtual machine has the tpm2.0 feature enabled.
-func (v *Validator) tpmEnabled(attestation vtpm.AttestationDocument, _ *attest.MachineState) error {
-	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-nitrotpm-support-on-ami.html
-	// 1. Get the vm's ami (from IdentiTyDocument.imageId)
-	// 2. Check the value of key "TpmSupport": {"Value": "v2.0"}"
-	ctx := context.Background()
-
-	idDocument := imds.InstanceIdentityDocument{}
-	err := json.Unmarshal(attestation.InstanceInfo, &idDocument)
+// sha512sum PEM-encodes a public key and calculates the SHA512 hash of the encoded key.
+func sha512sum(key crypto.PublicKey) ([64]byte, error) {
+	pub, err := x509.MarshalPKIXPublicKey(key)
 	if err != nil {
-		return err
+		return [64]byte{}, fmt.Errorf("marshalling public key: %w", err)
+	}
+
+	return sha512.Sum512(pub), nil
+}
+
+// Validate a given SNP report.
+type snpReportValidator interface {
+	validate(ctx context.Context, attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
```

Comment:
```suggestion
	validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
```
Since AWS and Azure don't share a common interface, you can drop the context from this method as its not used by the AWS code